src/ott/geometry/geodesic.py

```diff
@@ -0,0 +1,308 @@
+
+    return cls(laplacian, t=t, **kwargs)
+
+  def apply_kernel(
+      self,
+      scaling: jnp.ndarray,
+      eps: Optional[float] = None,
+      axis: int = 0,
+  ) -> jnp.ndarray:
+    r"""Apply :attr:`kernel_matrix` on positive scaling vector.
+
+    Args:
+    scaling: Scaling to apply the kernel to.
+    eps: passed for consistency, not used yet.
+    axis: passed for consistency, not used yet.
+
+    Returns:
+    Kernel applied to ``scaling``.
+    """
+
+    def compute_laplacian(adjacency_matrix: jnp.ndarray) -> jnp.ndarray:
+      """Compute the Laplacian matrix from the adjacency matrix.
+
+      Args:
+          adjacency_matrix: An (n, n) array representing the
+          adjacency matrix of a graph.
+
+      Returns:
+          An (n, n) array representing the Laplacian matrix.
+      """
+      degree_matrix = jnp.diag(jnp.sum(adjacency_matrix, axis=0))
+      return degree_matrix - adjacency_matrix
+
+    def compute_largest_eigenvalue(laplacian_matrix, k):
+      """Compute the largest eigenvalue of the Laplacian matrix.
+
+      Args:
+          laplacian_matrix: An (n, n) array representing the Laplacian matrix.
+          k: Number of eigenvalues/vectors to compute.
+
+      Returns:
+          The largest eigenvalue of the Laplacian matrix.
+      """
+      n = laplacian_matrix.shape[0]
+      initial_directions = jax.random.normal(jax.random.PRNGKey(0), (n, k))
+      # Convert the Laplacian matrix to a dense array
+      #laplacian_array = laplacian_matrix.toarray()
+      eigvals, _, _ = lobpcg_standard(laplacian_matrix, initial_directions, m=k)
+
+      return np.max(eigvals)
+
+    def rescale_laplacian(laplacian_matrix: jnp.ndarray) -> jnp.ndarray:
+      """Rescale the Laplacian matrix.
+
+      Args:
+          laplacian_matrix: An (n, n) array representing the Laplacian matrix.
+
+      Returns:
+          The rescaled Laplacian matrix.
+      """
+      largest_eigenvalue = compute_largest_eigenvalue(laplacian_matrix, k=1)
+      if largest_eigenvalue > 2:
+        rescaled_laplacian = laplacian_matrix.copy()
+        rescaled_laplacian /= largest_eigenvalue
+      return 2 * rescaled_laplacian
+
+    def define_scaled_laplacian(laplacian_matrix: jnp.ndarray) -> jnp.ndarray:
+      """Define the scaled Laplacian matrix.
+
+      Args:
+          laplacian_matrix: An (n, n) array representing the Laplacian matrix.
+
+      Returns:
+          The scaled Laplacian matrix.
+      """
+      n = laplacian_matrix.shape[0]
+      identity = jnp.eye(n)
+      return laplacian_matrix - identity
+
+    def chebyshev_coefficients(t: float, max_order: int) -> List[float]:
+      """Compute the coeffs of the Chebyshev pols approx using Bessel functs.
+
+      Args:
+          t: Time parameter.
+          max_order: Maximum order of the Chebyshev polynomial approximation.
+
+      Returns:
+          A list of coefficients.
+      """
+      return (2 * ive(jnp.arange(0, max_order + 1), -t)).tolist()
```

Comment:
This has to be done using https://jax.readthedocs.io/en/latest/_autosummary/jax.pure_callback.html, as you call the `scipy` implementation. Will go check the status of JAX implementation, but afaik it's not ready, hence the need for the pure callback.

Reply:
I have included this function into the `from_graph()` method, and it is looking now like this:

`
# Compute the coeffs of the Chebyshev pols approx using Bessel functs.
chebyshev_coeffs = (2 * ive(jnp.arange(0, geodesic_instance.order + 1), -geodesic_instance.t)).tolist()
`

Does it still need to be wrapped around the callback?